### PR TITLE
[WIP] Improvements to Disp3D and MNE Scan

### PR DESCRIPTION
**Pull request by LorenzE**
**Disp3D:**
- Restructured real-time workers for streaming and interpolating sensor data. 
  - Make use of timers for scheduling streaming data in order to get a more controllable and efficient streaming.
  - Make use of Qt's event system instead of subclassing QThread. 
  - Move Interpolation and visualization to different workers and threads. 
  - Introduce controller class to conrol both workers. 
  - All communciations with the two workers is now realized via queued Signal/Slots. 
  - Merge GPU and CPU workers from previous PR. 
- Fixed multiple calls of editor changes when changing control parameters (cancel distance, colormap, etc.) in the GUI/Tree view.
- Renamed some functions to be more consistent and descriptive.
- Moved CPU sensor data visualization to own child item as introduced in the last PR.
- Made use of the AbstractView class in all Disp3D related examples to get a better user experience.
- Added Escape button to exit full screen mode in View3D.
- Merged functionality of network threshold and other threshold use cases in order to be more consistent.
- Subclassed CpuSensorDataTreeItem and GpuSensorDataTreeItem from SensorDataTreeItem in order to reduce redundant code.
- Subclassed GPUInterpolationItem from AbstracMeshTreeItem to be consistent with other surfaces.
- Reduced the number of needed parameters when adding sensor level data to the Data3DTreeModel.
- Renamed most header guards due to wrong use of coding conventions.
- Improved memory/performance consumption when interpolating high resolution surfaces and high cancel distance.
  - Iterate over non-zero entries of sparse interpoaltion operator only. 
  - Make changes to how the QBuffers/GLBuffers are handled.
- Make use of new rt streaming structure and surface constrained interpolation methods for source level data/items. 
  - SCDC is now active for CPU Version when streaming source level data.
  - GPU version still needs some debugging for 32bit builds
- A simplified UML diagram of Disp3D should look something like this:

disp3d class diagram

**MNE Scan:**
- Corrected some spelling mistakes in most MNE Scan plugin headers.
- Got rid of TMSI SDK DLL check in the plugin.pro file in order to be consistent to other acquisition plugins. 
  - The check is still done during runtime.

### CODE REVIEW

File: libraries/disp3D/helpers/interpolation/interpolation.cpp [hunk 1]
```diff
@@ -112,63 +113,65 @@ QSharedPointer<SparseMatrix<double> > Interpolation::createInterpolationMat(cons
...
             if(!fiffInfo.bads.contains(s.ch_name)){
-                sensorLookup.insert (pProjectedSensors->at(idx));
+                sensorLookup.insert (vecProjectedSensors.at(idx));
             }
 
             idx++;
         }
     }
 
+    int counter = 0;
```
Comment (1DIce) [inline libraries/disp3D/helpers/interpolation/interpolation.cpp]: counter has no function.

File: libraries/disp3D/engine/model/workers/rtSensorData/rtsensordataworker.h [hunk 2]
```diff
@@ -262,90 +177,89 @@ class DISP3DSHARED_EXPORT RtSensorDataWorker : public QThread
...
-    * @param[in] info                 The fiff info including the new bad channels.
+    * @param[in] matInterpolationMatrix                 The new interpolation matrix.
     */
-    void updateBadChannels(const FIFFLIB::FiffInfo& info);
+    void setInterpolationMatrix(Eigen::SparseMatrix<float> matInterpolationMatrix);
 
     //=========================================================================================================
     /**
-    * Sets the running flag to false and waits for the worker to stop.
+    * Streams teh data.
```
Comment (1DIce) [inline libraries/disp3D/engine/model/workers/rtSensorData/rtsensordataworker.h]: Change spelling.

File: libraries/disp3D/engine/model/workers/rtSourceLoc/rtsourcedatacontroller.h [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+    * @param[in] iNumberVertsRight     The number of vertices for the right hemisphere.
+    */
+    void numberVerticesChanged(int iNumberVertsLeft,
+                               int iNumberVertsRight);
+
+    //=========================================================================================================
+    /**
+    * Emit this signal whenever the state to whether stream smoothed/interpolated or raw data changed.
+    *
+    * @param[in] info                 The fiff info including the new bad channels.
```
Comment (1DIce) [inline libraries/disp3D/engine/model/workers/rtSourceLoc/rtsourcedatacontroller.h]: Docu doesn't match.

File: libraries/disp3D/engine/model/workers/rtSourceLoc/rtsourcedataworker.cpp [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+
+//*************************************************************************************************************
+
+MatrixX3f RtSourceDataWorker::generateColorsFromSensorValues(const VectorXd &vecSensorValues,
+                                                             VisualizationInfo &visualizationInfoHemi)
+{
+    if(vecSensorValues.rows() != visualizationInfoHemi.matInterpolationMatrix.cols()) {
+        qDebug() << "RtSourceDataWorker::generateColorsFromSensorValues - Number of new vertex colors (" << vecSensorValues.rows() << ") do not match with previously set number of sensors (" << visualizationInfoHemi.matInterpolationMatrix.cols() << "). Returning...";
+        MatrixX3f matColor = visualizationInfoHemi.matOriginalVertColor;
```
Comment (1DIce) [inline libraries/disp3D/engine/model/workers/rtSourceLoc/rtsourcedataworker.cpp]: Saving the return value in matColor is unnecessary.

File: libraries/disp3D/engine/model/workers/rtSourceLoc/rtsourcedataworker.h [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+                                      double dThresholdX,
+                                      double dThreholdZ,
+                                      QRgb (*functionHandlerColorMap)(double v));
+
+    //=========================================================================================================
+    /**
+    * @brief generateColorsFromSensorValues        Produces the final color matrix that is to be emitted
+    *
+    * @param[in] vecSensorValues                   A vector of sensor signals
+    * @param[in/out] visualizationInfoHemi             The needed visualization info
```
Comment (1DIce) [inline libraries/disp3D/engine/model/workers/rtSourceLoc/rtsourcedataworker.h]: Indent visualization info.

File: libraries/disp3D/engine/model/workers/rtSourceLoc/rtsourceinterpolationmatworker.h [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        Eigen::MatrixX3f            matVertices;                    /**< Holds all vertex information. */
+
+        QList<FSLIB::Label>         lLabels;                        /**< The annotation labels. */
+        QList<int>                  vertNos;
+        QMap<qint32, qint32>        mapLabelIdSources;              /**< The mapped label ID to sources. */
+
+        QVector<qint32>             vecMappedSubset;                /**< Vector index position represents the id of the sensor and the qint in each cell is the vertex it is mapped to. */
+        QVector<QVector<int> >      vecNeighborVertices;            /**< The neighbor vertex information. */
+
+        double (*interpolationFunction) (double);                                       /**< Function that computes interpolation coefficients using the distance values. */
```
Comment (1DIce) [inline libraries/disp3D/engine/model/workers/rtSourceLoc/rtsourceinterpolationmatworker.h]: Please fix the indentation.

File: libraries/disp3D/helpers/geometryinfo/geometryinfo.h [hunk 7]
```diff
@@ -122,48 +122,41 @@ class DISP3DSHARED_EXPORT GeometryInfo
...
     //=========================================================================================================
     /**
-    * @brief scdc                  Calculates surface constrained distances on the mesh that is held by the passed MNEBemSurface
+    * @brief scdc                           Calculates surface constrained distances on a mesh.
     *
-    * @param bemSurface            The surface on which distances should be calculated
-    * @param pVecVertSubset        The subset of IDs for which the distances should be calculated
-    * @param dCancelDist           Distances higher than this are ignored, i.e. set to infinity
+    * @param[in] matVertices               The surface on which distances should be calculated.
+    * @param[in/out] vecNeighborVertices   The neighbor vertex information.
```
Comment (1DIce) [inline libraries/disp3D/helpers/geometryinfo/geometryinfo.h]: Delete the out in [in/out]

File: libraries/disp3D/helpers/geometryinfo/geometryinfo.h [hunk 8]
```diff
@@ -122,48 +122,41 @@ class DISP3DSHARED_EXPORT GeometryInfo
...
+    * @param[in] pVecVertSubset            The subset of IDs for which the distances should be calculated.
+    * @param[in] dCancelDist               Distances higher than this are ignored, i.e. set to infinity.
     *
-    * @return                      A double matrix. One column represents the distances for one vertex inside of the passed subset
+    * @return                               A double matrix. One column represents the distances for one vertex inside of the passed subset
     */
-    static Eigen::MatrixXd scdc(const MNELIB::MNEBemSurface &bemSurface,
+    static Eigen::MatrixXd scdc(const Eigen::MatrixX3f &matVertices,
+                                const QVector<QVector<int> > &vecNeighborVertices,
                                 QVector<qint32> &pVecVertSubset,
```
Comment (1DIce) [inline libraries/disp3D/helpers/geometryinfo/geometryinfo.h]: pVecVertSubset is missing a const.

File: libraries/disp3D/helpers/geometryinfo/geometryinfo.h [hunk 9]
```diff
@@ -175,38 +168,43 @@ class DISP3DSHARED_EXPORT GeometryInfo
...
     //=========================================================================================================
     /**
-    * @brief squared               Implemented for better readability only
-    * @param dBase                 Base double value
+    * @brief squared        Implemented for better readability only
     *
-    * @return                      Base squared
+    * @param[in] dBase     Base double value
+    *
+    * @return               Base squared
```
Comment (1DIce) [inline libraries/disp3D/helpers/geometryinfo/geometryinfo.h]: Indentation.

File: libraries/disp3D/helpers/geometryinfo/geometryinfo.h [hunk 10]
```diff
@@ -175,38 +168,43 @@ class DISP3DSHARED_EXPORT GeometryInfo
...
-    * @param itSensorBegin         The iterator that indicates the start of the wanted section of positions
-    * @param itSensorEnd           The iterator that indicates the end of the wanted section of positions
+    * @brief nearestNeighbor        Calculates the nearest vertex of an MNEmatVertices for each position between the two iterators
+    *
+    * @param[in] matVertices       The MNEmatVertices that holds the vertex information
+    * @param[in] itSensorBegin     The iterator that indicates the start of the wanted section of positions
+    * @param[in] itSensorEnd       The iterator that indicates the end of the wanted section of positions
     *
-    * @return                      A vector of nearest vertex IDs that corresponds to the subvector between the two iterators
+    * @return                       A vector of nearest vertex IDs that corresponds to the subvector between the two iterators
```
Comment (1DIce) [inline libraries/disp3D/helpers/geometryinfo/geometryinfo.h]: Indentation.

File: libraries/disp3D/helpers/interpolation/interpolation.h [hunk 11]
```diff
@@ -120,21 +120,19 @@ class DISP3DSHARED_EXPORT Interpolation
...
     *    -# if not: the values are calculated to give a total of 1 (a lot of values will stay 0, because they are too far away to influence) by using the above mentioned formula
     *
-    * @brief <i>createInterpolationMat</i>     Calculate weight matrix for later interpolation
-    *
-    * @param vecProjectedSensors               Vector of IDs of sensor vertices
-    * @param matDistanceTable                  Matrix that contains all needed distances
-    * @param interpolationFunction             Function that computes interpolation coefficients using the distance values
-    * @param dCancelDist                       Distances higher than this are ignored, i.e. the respective coefficients are set to zero
+    * @param[in] vecProjectedSensors           Vector of IDs of sensor vertices
+    * @param[in] matDistanceTable              atrix that contains all needed distances
```
Comment (1DIce) [inline libraries/disp3D/helpers/interpolation/interpolation.h]: Matrix is missing an M.

File: libraries/disp3D/helpers/geometryinfo/geometryinfo.h (hunk 8 above)
Comment (1DIce) [reply libraries/disp3D/helpers/geometryinfo/geometryinfo.h]: or make a copy if needed.

Comment (LorenzE) [reply libraries/disp3D/helpers/geometryinfo/geometryinfo.h]: a non const reference is needed because vecVertSubset gets changed if it is empty.

**Pull request merged**